cg.go

```diff
@@ -0,0 +1,162 @@
+package optimize
+
+import (
+	"math"
+
+	"github.com/gonum/floats"
+)
+
+const (
+	iterationRestartFactor = 6
+	angleRestartTreshold   = 0.9
+)
+
+type CGVariant interface {
```

Comment:
CGVariant computes the scaling parameter in the direction update formula, which is what differentiates various CG methods one from another. Therefore I like CGVariant, alternatively it could be called CGFormula. CGPreconditioner seems out of place here.